### PR TITLE
Fix for p4c Issue 2670

### PR DESCRIPTION
**Pull request by hesingh**
Add parenthesis to Operation_Binary to fix gcc compile error reported in https://github.com/p4lang/p4c/issues/2670

`expressionPrecedence` doesn't work for Operation_Binary yet that would also fix the gcc compile error.  

Another fix is to print, e.g., `0xffff0000` instead of its decimal format.

### CODE REVIEW

File: backends/ebpf/codeGen.cpp [hunk 1]
```diff
@@ -27,7 +27,9 @@ void CodeGenInspector::substitute(const IR::Parameter* p, const IR::Parameter* w
 { substitution.emplace(p, with); }
 
 bool CodeGenInspector::preorder(const IR::Constant* expression) {
-    builder->append(Util::toString(expression->value, 0, false));
+    // Until EBPF/UBPF applications set the IR::Constant base to hex vs.
+    // decimal all emits will use hex.
+    builder->append(Util::toString(expression->value, 0, false, 16));
```
Comment (mbudiu-vmw) [inline backends/ebpf/codeGen.cpp:32]: Why is this necessary?

File: backends/ebpf/codeGen.cpp [hunk 2]
```diff
@@ -49,19 +51,13 @@ bool CodeGenInspector::preorder(const IR::Declaration_Variable* decl) {
 
 bool CodeGenInspector::preorder(const IR::Operation_Binary* b) {
     widthCheck(b);
-    int prec = expressionPrecedence;
-    bool useParens = prec > b->getPrecedence();
-    if (useParens)
-        builder->append("(");
+    builder->append("(");
```
Comment (mbudiu-vmw) [inline backends/ebpf/codeGen.cpp:54]: Perhaps you can review the code and remove all uses of expressionPrecedence and remove the field too.

File: backends/ebpf/codeGen.cpp (hunk 1 above)
Comment (hesingh) [reply backends/ebpf/codeGen.cpp:32]: Otherwise `4294963200` is emitted instead of `0xfffff000` for 

`headers.ipv6.ip_version_traffic_class_and_flow_label & ~4294963200`

Comment (mbudiu-vmw) [reply backends/ebpf/codeGen.cpp:32]: Why is that incorrect?

Comment (hesingh) [reply backends/ebpf/codeGen.cpp:32]: Won't be incorrect.  A mask used with `&` is more readable as hex.  Any other constant would also be printed in hex, but that is fine too.  Note, even the bmv2 JSON format uses hex to emit any mask.

Comment (mbudiu-vmw) [reply backends/ebpf/codeGen.cpp:32]: This code is used for all constants, not just masks.
The right solution is to make sure that the constants created when expanding slices in the Lower pass actually use base 16.

File: backends/ebpf/codeGen.cpp (hunk 2 above)
Comment (hesingh) [reply backends/ebpf/codeGen.cpp:54]: If I remove all uses of expressionPrecedence, then all uses of its comparisons with object->getPrecedence() will be removed as well.  Can't have that or more gcc compiler failures crop up.  It's tricky.

**Comment by mbudiu-vmw** [comment]
The reference output you have checked-in for op_bin is different from the one computed by the compiler, and this is why the tests fail.

File: backends/ebpf/codeGen.cpp (hunk 2 above)
Comment (hesingh) [reply backends/ebpf/codeGen.cpp:53]: With the code changes of this PR to cogeGen.cpp, another test breaks.  It's as follows:

The P4 code below

`if (meta->_acl_outer21 == 1)`

is emitted as 

`if ((meta->_acl_outer21 == 1))`

With double parenthesis, gcc will cause an error.

In a subsequent checkin, I will add the line above and another line that can break gcc compiling with wrong C code generation.  The checkin will also add a fix that fixes all three issues - the `&` line of code has appropriate parenthesis, and so will the two lines of if-statements.

File: backends/ebpf/codeGen.cpp [hunk 3]
```diff
@@ -50,7 +51,10 @@ bool CodeGenInspector::preorder(const IR::Declaration_Variable* decl) {
 bool CodeGenInspector::preorder(const IR::Operation_Binary* b) {
     widthCheck(b);
     int prec = expressionPrecedence;
-    bool useParens = prec > b->getPrecedence();
+    bool useParens = false;
+    if ((prec > b->getPrecedence()) ||
```
Comment (mbudiu-vmw) [inline backends/ebpf/codeGen.cpp:55]: I think that what you want here is to check whether the parent is an if statement:
`useParents = getParent<IR::IfStatement>() == nullptr;`
This will use parentheses for everything but not for the expression in an if condition.

**Comment by mbudiu-vmw** [comment]
@hesingh : these tests failed because of the issue fixed by https://github.com/p4lang/p4c/pull/2676.
As soon as we merge that PR perhaps you can just rebase on master and we can merge this PR as well.

**Comment by hesingh** [comment]
> @hesingh : these tests failed because of the issue fixed by #2676.
> As soon as we merge that PR perhaps you can just rebase on master and we can merge this PR as well.

ok.

**Comment by mbudiu-vmw** [comment]
I have merged the other PR, please rebase this on master.

**Review by mbudiu-vmw — APPROVED** [review]
